auraed/src/runtime/cells/cell.rs

```diff
@@ -82,73 +80,83 @@ impl Cell {
 
     pub fn start_executable(
         &mut self,
-        exe_name: ExecutableName,
-        command: Command,
+        executable_name: ExecutableName,
+        command: String,
         args: Vec<String>,
         description: String,
     ) -> Result<()> {
+        // TODO: replace with try_insert when it becomes stable
         // Check if there was already an executable with the same name.
-        if self.executables.contains_key(&exe_name) {
+        if self.executables.contains_key(&executable_name) {
             return Err(CellError::ExecutableExists {
                 cell_name: self.name.clone(),
-                executable_name: exe_name,
+                executable_name,
             });
         }
 
+        let executable =
+            Executable::new(executable_name, command, args, description);
+        let executable_name = executable.name.clone();
+
+        // Ignoring return value as we've already assured ourselves that the key does not exist.
+        let _ = self.executables.insert(executable_name.clone(), executable);
+
         // Start the child process
-        let exe =
-            Executable::start(exe_name.clone(), command, args, description)
-                .map_err(|e| CellError::ExecutableError {
+        if let Some(executable) = self.executables.get_mut(&executable_name) {
+            let pid = executable.start().map_err(|e| {
+                CellError::FailedToStartExecutable {
                     cell_name: self.name.clone(),
+                    executable_name: executable.name.clone(),
+                    command: executable.command.clone(),
+                    args: executable.args.clone(),
                     source: e,
-                })?;
-
-        // TODO: If we start the exe above and fail the add below...bad...solution???
-
-        // Add the newly started child process to the cgroup
-        let exe_pid = exe.pid();
-        if let Err(e) = self.cgroup.add_task(exe.pid()) {
-            return Err(CellError::FailedToAddExecutable {
-                cell_name: self.name.clone(),
-                executable: exe,
-                source: e,
-            });
-        }
+                }
+            })?;
 
-        info!(
-            "Cells: cell_name={} executable_name={exe_name} spawn() -> pid={}",
-            self.name, exe_pid.pid
-        );
+            // TODO: We've inserted the executable into our in-memory cache, and started it,
```

Comment:
maybe we should only insert after we've started and added it to the cell?

Reply:
We had it as start -> add (to cgroup) -> insert. All three are technically fallible, but insert is the one we have the most control over, and can know will succeed because we check that the `HashMap` doesn't `contain_key` right before.

I was thinking about what we would want the state of the system/cache be if something fails. By doing the insert first, if anything else fails, at least we still have a reference to the executable and can decide what to do with it.

Depending on our order of execution, some of our choices seem to be:
1. (Insert last) executable doesn't start, we end the request -> no big deal.
2. (Insert last) Executable started and in wrong Cell (`self.cgroup.add_task` fails), and cache doesn't have a reference to it -> how do we heal from this state?
3. (Insert last) Executable started and in correct Cell, but cache doesn't have a reference to it -> unlikely; insert is going to succeed
4.  (Insert first) we always have a reference no matter the outcome of the other calls -> if anything fails after, we have the most options
5. (Insert middle) executable starts, and we have a reference to it, but maybe in the wrong cell -> probably the same benefits of 4?